### ISSUE TITLE
Swagger Spec is outdated and needs reusable model components

### ISSUE THREAD

**Issue by ff137** [odota/core#2655]
This issue became quite lengthy, because I give you instructions for replicating my findings ... 
TL;DR: the spec needs fixing, and here's how you can generate an OpenAPI client to prove it.

---

`routes/spec.js` is outdated (Swagger 2.0 < OpenApi 3.x), and it contains some incorrect or inconsistent typing in its responses.

This can be observed by generating an OpenApi client from the spec.

Conveniently, I've documented the steps to do this (or, ChatGPT mostly did it for me). 
If you want to generate the python openapi client and validate it yourself:

---

### Prerequisites

  Before getting started, you need to have the following tools installed:
  
  - [Node.js and npm](https://nodejs.org/en/download/)
  - [OpenAPI Generator CLI](https://github.com/OpenAPITools/openapi-generator)
  
  You can install the OpenAPI Generator CLI using npm:
  
```sh
npm install @openapitools/openapi-generator-cli -g
```

As of April 2023, the OpenDota API has a Swagger specification version of 2.0. The `openapi-generator-cli` requires a Swagger spec of at least 3.0, so include installation of the following convertor:

```sh
npm install -g swagger2openapi
```

To generate Python classes from the OpenDota API specification, here I've documented instructions:

  1. Download the OpenDota API specification JSON file from the following URL: `https://api.opendota.com/api`. Save it as `opendota-api-spec.json` in the `opendota_client/` folder.

```sh
# Download the OpenDota API specification JSON file
curl https://api.opendota.com/api -o opendota_client/opendota-api-spec.json
```
  
  3. Convert the `opendota-api-spec.json` file (Swagger 2.0 format) to OpenAPI 3.x format:
  
```sh
swagger2openapi opendota_client/opendota-api-spec.json -o opendota_client/opendota-api-spec-openapi3.json
```
  
This command will output a new JSON file opendota-api-spec-openapi3.json in OpenAPI 3.x format.
  
  4. The newly generated spec with OpenAPI 3.x will throw some warnings. Ignore it for now. (I wrote a python script to fix warnings on my side)
  
  ### Generating Python Classes
  
  Navigate and run the `openapi-generator-cli` command on our newly updated json file, to generate the Python client classes:
  
```sh
openapi-generator-cli generate -i opendota_client/opendota-api-spec-openapi3-updated.json -g python -o opendota_client
```
 
---

Done, you should be able to generate your openapi_client with those steps.

Here is some sample python code demonstrating how to get hero_ids, with this openapi client, and to get benchmarks - for example (look at apis.tags for all the other routes):

```py
from opendota_client.openapi_client import Configuration, ApiClient, ApiException
from pprint import pprint
from opendota_client.openapi_client.apis.tags import (
    benchmarks_api,
)

# Defining the host is optional and defaults to https://api.opendota.com/api
# See configuration.py for a list of all supported configuration parameters.
configuration = Configuration(
    host = "https://api.opendota.com/api"
)

# Enter a context with an instance of the API client
with ApiClient(configuration) as api_client:
    # Create an instance of the API class
    
    # HeroesApi
    heroes_instance = heroes_api.HeroesApi(api_client)
    try:
        pprint("heroes")
        heroes_response = heroes_instance.get_heroes().body
        pprint(heroes_response)
    except ApiException as e:
        print("Exception when calling HeroesApi->get_heroes: %s\n" % e)
    
    hero_ids = [hero['id'] for hero in heroes_response]
    print("hero_ids = ", hero_ids)

    # BenchmarksApi
    benchmarks_instance = benchmarks_api.BenchmarksApi(api_client)
    try:
        pprint("benchmarks")
        for h in hero_ids:
            hero_id = {"hero_id": f"{h}"}
            api_response = benchmarks_instance.get_benchmarks(hero_id)
            pprint(api_response)
    except ApiException as e:
        print("Exception when calling BenchmarksApi->get_benchmarks: %s\n" % e)
```
Your method names will differ due to the warnings from earlier - my script cleaned up the auto-generated operationIds to be different.

In doing this, you will discover that the spec is incompatible with the server responses. (for example, some benchmarks return strings when integers are expected, and getMatches has many fields that need to be nullable). This is some strange behavior though, because to the best of my knowledge, from this code block at the end of `routes/api.js`, it looks like the server responses are directly inferred from the js spec ...
```js
// API endpoints
Object.keys(spec.paths).forEach((path) => {
  Object.keys(spec.paths[path]).forEach((verb) => {
    const { route, func } = spec.paths[path][verb];
    api[verb](route(), func);
  });
});
```
so it's not clear why the spec says one thing but the api returns another? Does the type inconsistency comes from how data is queried from storage? 🤷‍♂️ Too much code and search isn't helping me.

I've already created a PR for fixing some of the more clear errors within the spec: https://github.com/odota/core/pull/2654

But this is clearly a bigger task than I expected. Improving the spec, and updating to OpenAPI 3.x will be very beneficial.

Additionally, I notice that the auto-generated openapi_client has empty models folders. This will be a huge boon to the project if these can be auto-generated by OpenAPI, and will offer much better maintainability to the spec itself ... several responses have different type definitions for fields common in players / matches / records / etc, and spec components will help standardise this.

Model components would look something like this at the top of the `spec.js`:
```yaml
components:
  schemas:
    ModelName:
      type: object
      properties:
        field1:
          type: string
        field2:
          type: integer
```

Any feedback will be appreciated! I like reverse engineering things, but this becomes tough to fix when it's unclear what other services are affects by spec changes, and how to test for it. 

I'm interested in writing my own custom api client from the spec, so I can help out a lot with these improvements, but guidance I will need! So let's solve this together :-)

**Comment by ff137** [comment]
Resolved since this PR #2656 upgraded spec to OpenAPI with component definitions.
